format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressorStream.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.charset.fixedlength;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.io.compress.DecompressorStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * DecompressorStream implementation for the FixedLengthCharsetTransformingDecompressor.
+ * <p>
+ * This can be used to convert input streams containing bytes for fixed length charsets into UTF-8 bytes.
+ */
+public class FixedLengthCharsetTransformingDecompressorStream extends DecompressorStream {
```

Comment:
What does this inheritance give you? You are only using it to wrap into `TransformingCompressionInputStream`. Would it make sense to move logic from here to `TransformingCompressionInputStream` and use `DecompressorStream` from `TransformingCompressionInputStream`. Having one less layer can make position handling easier.

Reply:
Generally I am not sure if using DecompressorStream simplifies things or makes it harder. BZip2CompressionInputStream does not use it and it looks very clean and small.

Reply:
The DecompressorStream gives me a place where I can check for partition boundaries and align the reads with the partition boundary. This ensures the Line Record Reader is able to read all the lines in a partition: https://github.com/cdapio/hydrator-plugins/blob/125413aeb91761443be156c954426dc91869793a/format-common/src/main/java/io/cdap/plugin/format/charset/CharsetTransformingLineRecordReader.java#L159

Reply:
You are using `SplitCompressionInputStream` in there. In it you are building upon `FixedLengthCharsetTransformingDecompressorStream` that extends `DecompressorStream` My question is if it simpler to implement SplitCompressionInputStream directly without additional layers, especially extending `DecompressorStream` in a non-trivial fragile way. The one example implementation of `SplitCompressionInputStream` i hadoop is `BZip2CompressionInputStream` that do not use DecompressorStream underneath, but works with BZIP algorithm directly.

Reply:
The DecompressorStream gives us an avenue to test the partition boundaries in https://github.com/cdapio/hydrator-plugins/blob/29aaaf55010acbc6fb19c775a3f64fbfb2c7e3df/format-common/src/test/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressorStreamTest.java.

I do not think the encapsulation/additional layers really pose a problem for us. I tried to keep the implementation as similar as possible to the original LineRecordReader class I copied for simplicity/maintainability in the future.

Reply:
Changes for that test is actually one of the biggest indicators of structure issues. Having a method like `getCompressedDataSuper` is a big smell of structural problems. Let's to sync up on this partition boundary synchronization, with a good integration test that shows the problem we are trying to solve.

Reply:
The purpose of the test is to demonstrate that the super method exceeds the partition boundary, which poses a problem when reading lines.

I will come up with a scenario to demonstrate this issue tomorrow.